assets/js/broken-links.js

```diff
@@ -0,0 +1,198 @@
+class BrokenLinks {
+  constructor(rootUrl) {
+    this.rootUrl = rootUrl;
+    this.errors = [];
+    this.links = {};
+    this.prevLink = {};
+    this.options = {
+      mode: 'no-cors',
+      redirect: 'follow',
+    };
+    this.end = false;
+    this.duration = 3000;
+
+    setTimeout(() => {
+      clearInterval(this.id);
+    }, 600000);
+
+    this.start();
+  }
+
+  async start() {
+    await this.getUrl(this.rootUrl);
+    this.id = setInterval(() => {
+      this.handleIntervall();
+    }, this.duration);
+  }
+
+  handleIntervall() {
+    const reductionFactor = 1.5;
+    const lowerLimit = 500;
+    const upperLimit = this.duration;
+
+    if (this.end) {
+      if (this.duration < lowerLimit) {
+        this.displayResult();
+        clearInterval(this.id);
+      }
+      this.duration = Math.floor(this.duration / reductionFactor);
+    } else {
+      this.duration = upperLimit;
+    }
+  }
+
+  displayResult() {
+    const container = document.getElementById('broken-links-table');
+    if (!container) return console.error('No Container found');
+
+    const table = document.createElement('table');
+    table.classList.add('table');
+    container.appendChild(table);
+
+    let head = table.createTHead();
+    let row = head.insertRow(0);
+    let leftCell = row.insertCell(0);
+    let rightCell = row.insertCell(1);
+
+    let leftCellText = 'Broken Links';
+    let rightCellText = 'Previous Links';
+    leftCell.innerHTML = leftCellText;
+    rightCell.innerHTML = rightCellText;
+
+    let body = table.createTBody();
+
+    for (let i = 0; i < this.errors.length; i++) {
+      const url = this.errors[i].url;
+      const prevUrl = this.prevLink[url];
+
+      const tableRow = body.insertRow(i);
+      const tableCellBrokenLink = tableRow.insertCell(0);
+      const tableCellPrevLink = tableRow.insertCell(1);
+
+      const brokenLinkText = document.createTextNode(url);
+      const previousLinkText = document.createTextNode(prevUrl);
+      tableCellBrokenLink.appendChild(brokenLinkText);
+      tableCellPrevLink.appendChild(previousLinkText);
+    }
+  }
+
+  hasDuplicates(url) {
+    return this.errors.some((x) => x['url'] === url);
+  }
+
+  getUrl(url, previousUrl) {
+    this.end = false;
+
+    this.prevLink[url] = previousUrl;
+    if (this.hasDuplicates(url)) return;
+
+    return new Promise((resolve) => {
+      const external = this.isExternal(url);
+
+      fetch(url, this.options)
+        .then((response) => {
+          this.handleResponse(response, external);
+          this.end = true;
```

Comment:
zwischen dem this.end = true und dem this.end = false passiert aber noch jede menge oder nicht? glaube das this.end = false muss früher, schau dir mal den callstack an.